Auto-detect existing credentials for new features

faisal-alvi opened the pull request:
<!--
Filling out this template is required.  Any PR that does not include enough information to be reviewed may be closed at a maintainers' discretion.  All new code requires documentation and tests to ensure against regressions.
-->

### Description of the Change

This PR implements auto-detection and reuse of existing service provider credentials when enabling new features, addressing the UX friction of having to re-enter the same API credentials multiple times.

- Added a **CredentialReuse Helper Class**
- Added new **REST API Endpoints**
- Added **React Modal Component**
- Integrated a requested feature

## Modal

![Image](https://github.com/user-attachments/assets/07aec3f6-6214-488a-9f79-1393416b4d87)

🔍 
![image](https://github.com/user-attachments/assets/f5a4208a-4d87-4c58-92c3-1fa5c5bb5b8b)

<!-- Enter any applicable Issue number(s) here that will be closed/resolved by this PR. -->
Closes #936

### How to test the Change

1. **Setup Initial Provider:**
   - Go to ClassifAI settings
   - Configure OpenAI for "Title Generation" with valid API key
   - Save settings

2. **Test from Feature List page:**
   - Navigate to ClassifAI main settings page
   - Click "Enable" on "Image Generation" feature
   - Modal should appear asking: "You're trying to enable Image Generation and are already using OpenAI for Title Generation. Would you like to reuse the same OpenAI credentials?"
   - Test selecting "Reuse Credentials" - credentials should be copied and feature enabled
   - Verify Image Generation now shows OpenAI as configured

3. **Test from Feature's Individual page:**
   - Navigate to a different unconfigured feature (e.g., "Excerpt Generation")
   - Click the feature enable toggle
   - Modal should appear again with reuse option
   - Test selecting "Skip" - feature should enable without copying credentials

4. **Test "Don't Ask Again":**
   - Check "Don't ask me again" checkbox before selecting option
   - Enable another feature - modal should not appear
   - Clear localStorage and test that modal appears again

5. **Test Multiple Providers:**
   - Configure Azure AI for another feature
   - Enable a new feature that supports both OpenAI and Azure
   - Modal should show radio buttons for both providers

6. Test Don't ask again option.

### Changelog Entry

> Added - Auto-detection and Reuse of existing service provider credentials when enabling new features

### Credits
<!-- Please list any and all contributors on this PR so that they can be added to this projects CREDITS.md file. -->
Props @jeffpaul @dkotter @faisal-alvi 

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you are unsure about any of these, please ask for clarification.  We are here to help! -->
- [x] I agree to follow this project's [**Code of Conduct**](https://github.com/10up/.github/blob/trunk/CODE_OF_CONDUCT.md).
- [ ] I have updated the documentation accordingly.
- [ ] I have added [Critical Flows, Test Cases, and/or End-to-End Tests](https://10up.github.io/Open-Source-Best-Practices/testing/) to cover my change.
- [ ] All new and existing tests pass.

faisal-alvi committed:
Add CredentialReuse helper class for provider credential management

faisal-alvi committed:
Add REST API endpoints for credential reuse functionality

faisal-alvi committed:
Add React modal component for credential reuse selection

faisal-alvi committed:
Integrate credential reuse with individual feature settings pages

faisal-alvi committed:
Integrate credential reuse with feature list dashboard

faisal-alvi committed:
Add CSS styles for credential reuse modal interface

faisal-alvi committed:
Fix Linting Errors

faisal-alvi committed:
Merge branch 'develop' into enhance/936

faisal-alvi committed:
PHPCS errors fix

faisal-alvi committed:
A few more eslint fixes

faisal-alvi committed:
A small spacing fix

faisal-alvi committed:
Eslint errors fixes

faisal-alvi committed:
Add commands to manage credential reuse modal visibility in tests, trigger modal disable when feature gets activated

dkotter commented on includes/Classifai/Helpers/CredentialReuse.php:
```diff
@@ -0,0 +1,210 @@
+	public static function get_provider_display_name( string $provider_id ): string {
+		$provider_names = [
+			'openai_chatgpt'        => __( 'OpenAI ChatGPT', 'classifai' ),
+			'openai_dalle'          => __( 'OpenAI DALL-E', 'classifai' ),
```
```suggestion
			'openai_dalle'          => __( 'OpenAI Images', 'classifai' ),
```

dkotter commented on src/js/settings/components/credential-reuse/index.js:
```diff
@@ -0,0 +1,257 @@
+						{ sprintf(
+							/* translators: %s: Feature label */
+							__(
+								'We found existing provider credentials that can be used with %s. Would you like to reuse them?',
```
```suggestion
								'We found existing Provider credentials that can be used with %s. Would you like to reuse them?',
```

dkotter commented on src/js/settings/components/credential-reuse/index.js:
```diff
@@ -0,0 +1,257 @@
+								<label // eslint-disable-line jsx-a11y/label-has-associated-control
+									aria-label={ sprintf(
+										/* translators: %s: Provider name */
+										__( 'Select %s provider', 'classifai' ),
```
```suggestion
										__( 'Select %s Provider', 'classifai' ),
```

dkotter commented on src/js/settings/components/credential-reuse/index.js:
```diff
@@ -0,0 +1,257 @@
+						checked={ dontAskAgain }
+						onChange={ setDontAskAgain }
+						help={ __(
+							'You can always configure credentials manually in the feature settings.',
```
```suggestion
							'You can always configure credentials manually in the Feature settings.',
```

dkotter commented on src/js/settings/components/feature-settings/enable-feature.js:
```diff
@@ -33,20 +102,26 @@ export const EnableToggleControl = ( { children } ) => {
-				__nextHasNoMarginBottom
+		<>
+			<SettingsRow
+				label={ __( 'Enable feature', 'classifai' ) }
```
```suggestion
				label={ __( 'Enable Feature', 'classifai' ) }
```

dkotter commented on tests/cypress/support/commands.js:
```diff
@@ -852,3 +853,23 @@ Cypress.Commands.add( 'classicCreateProduct', ( { title, content } ) => {
+ * Enable credential reuse modal (for testing the modal itself).
+ * This allows the modal to appear when enabling features.
+ */
+Cypress.Commands.add( 'enableCredentialReuseModal', () => {
```
Can we add some tests around this modal?

dkotter reviewed (commented):
Nice work here!

Did some quick testing and had a couple thoughts/questions.

If I enable a Feature from the main settings page, I immediately get the modal asking me if I want to re-use credentials. If I decide yes, the page refreshes and I stay on that main settings page.

I wonder if we should wait until someone clicks in to configure the settings to show the modal? It was a little jarring to me to enable it and immediately get the modal, though maybe that's just me.

At the very least, I think if we show the modal on that main screen, if they accept the credentials, we should send them to the settings screen for that Feature so it's clear they've configured things.

Another thing I noticed is we don't consider credentials that use different endpoints. For instance, an OpenAI API key can be used to get chat completions, generate an image, generate embeddings, generate audio, etc. But if I have an OpenAI API key set in Title Generation, if I enable Moderation, it won't prompt me to re-use credentials, even though those credentials are perfectly valid. I think we need some sort of mapping here between these different Features that tells us credentials from one are valid for another, even if those Features use different endpoints.

faisal-alvi committed:
Apply suggestions from code review

Co-authored-by: Darin Kotter <darin.kotter@gmail.com>

faisal-alvi committed:
Navigate to feature detail on credential reuse

faisal-alvi committed:
Improve modal visibility conditions based on loading state

faisal-alvi committed:
Enhanced credential reuse logic to consider group compatibility

faisal-alvi committed:
Add command to disable feature if already enabled in tests

faisal-alvi committed:
Add tests for Credential Reuse Modal behavior and interactions

faisal-alvi committed:
phpcs fixes,
FILE: CredentialReuse.php
-----------------------------------------------------------------------------------------------
FOUND 5 ERRORS AND 4 WARNINGS AFFECTING 9 LINES
-----------------------------------------------------------------------------------------------
  36 | WARNING | Array double arrow not aligned correctly; expected 2 space(s) between "'azure'" and double arrow, but found 1. (WordPress.Arrays.MultipleStatementAlignment.DoubleArrowNotAligned)
 140 | ERROR | Whitespace found at end of line (Squiz.WhiteSpace.SuperfluousWhitespace.EndLine)
 146 | ERROR | Whitespace found at end of line (Squiz.WhiteSpace.SuperfluousWhitespace.EndLine)
 161 | ERROR | Whitespace found at end of line (Squiz.WhiteSpace.SuperfluousWhitespace.EndLine)
 166 | ERROR | Whitespace found at end of line (Squiz.WhiteSpace.SuperfluousWhitespace.EndLine)
 173 | ERROR | Whitespace found at end of line (Squiz.WhiteSpace.SuperfluousWhitespace.EndLine)
 207 | WARNING | Array double arrow not aligned correctly; expected 9 space(s) between "'feature_id'" and double arrow, but found 4. (WordPress.Arrays.MultipleStatementAlignment.DoubleArrowNotAligned)
 208 | WARNING | Array double arrow not aligned correctly; expected 6 space(s) between "'feature_label'" and double arrow, but found 1. (WordPress.Arrays.MultipleStatementAlignment.DoubleArrowNotAligned)
 209 | WARNING | Array double arrow not aligned correctly; expected 8 space(s) between "'credentials'" and double arrow, but found 3. (WordPress.Arrays.MultipleStatementAlignment.DoubleArrowNotAligned)
-----------------------------------------------------------------------------------------------

faisal-alvi committed:
jslint fixes,
/home/runner/work/classifai/classifai/src/js/settings/components/credential-reuse/index.js
Error:   127:8  error    Variables should not be assigned until just prior its first reference. An early return statement may leave this variable unused  @wordpress/no-unused-vars-before-return

/home/runner/work/classifai/classifai/src/js/settings/components/service-settings/index.js
Error:   156:36  error    'providerId' is defined but never used      no-unused-vars

faisal-alvi committed:
Use enableFeature command with parameter for disabling modal

faisal-alvi committed:
Merge branch 'develop' into enhance/936 and resolve conflicts

faisal-alvi committed:
Try using different endpoint format
